doublerachet.go

```diff
@@ -0,0 +1,290 @@
+This function takes the following parameters:
+
+secret:    		Shared secret that was agreed with the initiator side
+remotePubKey 	The initiator side public key.
+
+This function creates a session and stores its state in the SessionStore
+Following this operation the caller can imediately use Encrypt/Decrypt by
+providing the sessionID as identifier.
+*/
+func NewSessionWithRemoteKey(secret, remotePubKey string) (string, error) {
+	sID, err := generateSessionID()
+	if err != nil {
+		return "", err
+	}
+
+	skBytes, err := hex.DecodeString(secret)
+	if err != nil {
+		return "", err
+	}
+
+	pubKeyBytes, err := hex.DecodeString(remotePubKey)
+	if err != nil {
+		return "", err
+	}
+
+	_, err = doubleratchet.NewWithRemoteKey([]byte(sID), toKey(skBytes), toKey(pubKeyBytes), &BoltDBSessionStorage{})
+	if err != nil {
+		return "", err
+	}
+
+	return sID, nil
+}
+
+/*
+Encrypt is used to encrypt a message providing a sessionID and a message to encrypt
+This function loads the session from the session store and use it to encrypt the message.
+This way the user is free of managing sessions state and only need to provide the sessionID.
+*/
+func RatchetEncrypt(sessionID, message string) (string, error) {
+	session, err := doubleratchet.Load([]byte(sessionID), &BoltDBSessionStorage{})
+	if err != nil {
+		return "", err
+	}
+	msg, err := session.RatchetEncrypt([]byte(message), nil)
+	if err != nil {
+		return "", err
+	}
+
+	rawMessage, err := json.Marshal(msg)
+	if err != nil {
+		return "", err
+	}
+	return string(rawMessage), nil
+}
+
+/*
+Decrypt is used to decrypt a message providing a sessionID and a message to decrypt
+This function loads the session from the session store and use it to decrypt the message.
+This way the user is free of managing sessions state and only need to provide the sessionID.
+*/
+func RatchetDecrypt(sessionID, message string) (string, error) {
+
+	var msg doubleratchet.Message
+	if err := json.Unmarshal([]byte(message), &msg); err != nil {
+		return "", err
+	}
+
+	session, err := doubleratchet.Load([]byte(sessionID), &BoltDBSessionStorage{})
+	if err != nil {
+		return "", err
+	}
+	decrypted, err := session.RatchetDecrypt(msg, nil)
+	if err != nil {
+		return "", err
+	}
+	return string(decrypted), nil
+}
+
+//DHPair is a key pair structure that implements the doubleratchet.DHPair interface
+type DHPair struct {
+	privateKey [32]byte
+	publicKey  [32]byte
+}
+
+//PrivateKey is part of the doubleratchets.DHPair interaface
+func (p DHPair) PrivateKey() doubleratchet.Key {
+	return p.privateKey
+}
+
+//PublicKey is part of the doubleratchets.DHPair interaface
+func (p DHPair) PublicKey() doubleratchet.Key {
+	return p.publicKey
+}
+
+//BoltDBSessionStorage is a tructure that implements the SessionStore interface.
+//It uses boltdb to save sessions
+type BoltDBSessionStorage struct{}
+
+// Save the session to the session store
+func (s *BoltDBSessionStorage) Save(id []byte, state *doubleratchet.State) error {
```

Comment:
I might be missing something but I only see this called in the tests. Because of the changing keys are we not supposed to save state after every message?

Reply:
This is called automatically by the library: 
https://github.com/status-im/doubleratchet/blob/develop/session.go#L97
https://github.com/status-im/doubleratchet/blob/develop/session.go#L118